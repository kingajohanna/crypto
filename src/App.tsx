import "react-native-gesture-handler";
import { BottomNavigator } from "@navigation/BottomNavigator";
import React from "react";

const App = () => {
    return <BottomNavigator />;
};

export default App;
