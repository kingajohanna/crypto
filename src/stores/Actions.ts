import { KeyValue, MarketData } from "@constants/DataTypes";
import { store } from "@stores/store";

export const LOGIN = "LOGIN";
export const LOGOUT = "LOGOUT";

export const FETCHMARKET = "FETCHMARKET";
export const FETCHFAVS = "FETCHFAVS";
export const FETCHCOINS = "FETCHCOINS";

export const FETCHOWNEDCOINS = "FETCHOWNEDCOINS";

export const SETERROR = "SETERROR";

const userLogin = (id: string, email: string, createdAt: string, name?: string, photoURL?: string) => ({
    type: LOGIN,
    id,
    email,
    createdAt,
    name,
    photoURL,
});

const userLogout = () => ({
    type: LOGOUT,
});

const fetchMarket = (marketCoins: MarketData[]) => ({
    type: FETCHMARKET,
    marketCoins,
});

const fetchFavs = (favs: MarketData[]) => ({
    type: FETCHFAVS,
    favs,
});

const fetchCoins = (coins: KeyValue[]) => ({
    type: FETCHCOINS,
    coins,
});

const fetchOwnedCoins = (ownedCoins: KeyValue[]) => ({
    type: FETCHOWNEDCOINS,
    ownedCoins,
});

const setError = (error: String) => ({
    type: SETERROR,
    error,
});

export const loginAction = (uid: string, email: string, createdAt: string, name?: string, photoURL?: string) => {
    try {
        store.dispatch(userLogin(uid, email, createdAt, name, photoURL));
    } catch (error) {
        console.log(error);
    }
};

export const logoutAction = () => {
    try {
        store.dispatch(userLogout());
        store.dispatch(fetchMarket([]));
        store.dispatch(fetchFavs([]));
    } catch (error) {
        console.log(error);
    }
};

export const fetchMarketAction = (marketCoins: MarketData[]) => {
    try {
        store.dispatch(fetchMarket(marketCoins));
    } catch (error) {
        console.log(error);
    }
};

export const fetchFavsAction = (favs: MarketData[]) => {
    try {
        store.dispatch(fetchFavs(favs));
    } catch (error) {
        console.log(error);
    }
};

export const fetchCoinsAction = (coins: KeyValue[]) => {
    try {
        store.dispatch(fetchCoins(coins));
    } catch (error) {
        console.log(error);
    }
};

export const fetchOwnedCoinsAction = (ownedCoins: KeyValue[]) => {
    try {
        store.dispatch(fetchOwnedCoins(ownedCoins));
    } catch (error) {
        console.log(error);
    }
};

export const setErrorAction = (error: string) => {
    try {
        store.dispatch(setError(error));
    } catch (error) {
        console.log(error);
    }
};
