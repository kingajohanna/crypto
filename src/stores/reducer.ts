import { Coins, KeyValue, MarketData } from "@constants/DataTypes";
import { FETCHCOINS, FETCHFAVS, FETCHMARKET, FETCHOWNEDCOINS, LOGIN, LOGOUT, SETAUTHERROR, SETERROR } from "@stores/Actions";
import { AnyAction } from "redux";

const initialUser = {
    id: "",
    email: "",
    createdAt: "",
    name: "",
    photoURL: "",
    isLoggedIn: false,
};

export type Crypto = {
    marketCoins: MarketData[];
    favs: MarketData[];
    ownedCoins: Coins[];
    coins: KeyValue[];
};

const initialCrypto = {
    marketCoins: [],
    favs: [],
    coins: [],
    ownedCoins: [],
};

export type Error = {
    error: string;
    authError: string;
};

const initialError = {
    error: "",
    authError: "",
};

export const authReducer = (state = initialUser, action: AnyAction) => {
    switch (action.type) {
        case LOGIN:
            return {
                ...state,
                id: action.id,
                email: action.email,
                createdAt: action.createdAt,
                name: action.name,
                photoURL: action.photoURL,
                isLoggedIn: true,
            };
        case LOGOUT:
            return initialUser;
        default:
            return state;
    }
};

export const cryptoReducer = (state = initialCrypto, action: AnyAction) => {
    switch (action.type) {
        case FETCHMARKET:
            return {
                ...state,
                marketCoins: action.marketCoins.slice(),
            };
        case FETCHFAVS:
            return {
                ...state,
                favs: action.favs.slice(),
            };
        case FETCHCOINS:
            return {
                ...state,
                coins: action.coins.slice(),
            };
        case FETCHOWNEDCOINS:
            return {
                ...state,
                ownedCoins: action.ownedCoins,
            };
        default:
            return state;
    }
};

export const errorReducer = (state = initialError, action: AnyAction) => {
    switch (action.type) {
        case SETERROR:
            return {
                ...state,
                error: action.error.slice(),
            };
        case SETAUTHERROR:
            return {
                ...state,
                authError: action.error.slice(),
            };
        default:
            return state;
    }
};
