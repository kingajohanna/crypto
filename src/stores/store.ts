import { authReducer, cryptoReducer, errorReducer } from "@stores/reducer";
import { combineReducers, createStore } from "redux";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { persistReducer } from "redux-persist";
import persistStore from "redux-persist/es/persistStore";

const rootReducer = combineReducers({
    user: authReducer,
    crypto: cryptoReducer,
    error: errorReducer,
});

const persistConfig = {
    key: "root",
    storage: AsyncStorage,
    whitelist: ["user", "crypto"],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export type RootState = ReturnType<typeof rootReducer>;

export const store = createStore(persistedReducer);
export const persistor = persistStore(store);
