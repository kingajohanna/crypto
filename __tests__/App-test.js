import React from "react";
import { act } from "react-test-renderer";
import { store } from "../src/stores/store";
import { render } from "@testing-library/react-native";
import App from "../src/App";
import { Provider } from "react-redux";

describe("App", () => {
    it("renders app stack", async () => {
        const result = render(
            <Provider store={store}>
                <App />
            </Provider>,
        );
        await act(async () => {
            expect(result).toMatchSnapshot();
        });
    });
});
