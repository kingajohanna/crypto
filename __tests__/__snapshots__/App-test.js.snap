// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders app stack 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#0B0C10",
      },
    ]
  }
/>
`;
